### PR TITLE
feat: Modules for implementing visitor nodes.

### PR DESCRIPTION
**Pull request by damencho**
Still WIP, uses visitor nodes prosodies where we create the main participants and forward the visitors to watch. Used for huge conferences.

### CODE REVIEW

File: resources/prosody-plugins/mod_fmuc.lua [hunk 1]
```diff
@@ -0,0 +1,134 @@
+--- activate under main muc component
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_fmuc.lua:1]: These are instructions setting up the system as a whole, not just mod_fmuc. Maybe they should go to a separate md file?

File: resources/prosody-plugins/mod_fmuc.lua [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+local muc_domain_prefix = module:get_option_string('muc_mapper_domain_prefix', 'conference');
+local main_domain = string.gsub(module.host, muc_domain_prefix..'.', '');
+
+local function get_focus_occupant(room)
+    local focus_occupant = room._data.focus_occupant;
+
+    if focus_occupant then
+        return focus_occupant;
+    end
+    if not focus_occupant then
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_fmuc.lua:28]: Is this redundant given that the case above returns?

File: resources/prosody-plugins/mod_fmuc.lua [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        for _, n_occupant in room:each_occupant() do
+            if jid.node(n_occupant.jid) == 'focus' then
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_fmuc.lua:30]: We recognize jicofo based on the nickname. Is it possible for users to login with this nickname?

File: resources/prosody-plugins/mod_fmuc.lua [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                     :tag('item', {
+            affiliation = room:get_affiliation(occupant.bare_jid) or 'none';
+            role = 'none';
+            nick = event.nick;
+            jid = occupant.bare_jid }):up()
+                :up();
+        room:route_stanza(pr);
+    end
+
+    if not room.destroying then
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_fmuc.lua:74]: Does `destroying` come from the MUC module itself, or is it something custom?

File: resources/prosody-plugins/mod_fmuc.lua [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        if room.xxl_destroy_timer then
+            room.xxl_destroy_timer:stop();
+        end
+
+        room.xxl_destroy_timer = module:add_timer(15, function()
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_fmuc.lua:79]: Has this come up in profiling? You create a timer every time someone leaves, and the task loops over all participants. The loop can be easily optimized to shortcurcuit if `main_count > 0 && visitors_count > 0`

File: resources/prosody-plugins/mod_xxl_conference.lua [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+local main_muc_component_config = module:get_option_string('main_muc');
+if main_muc_component_config == nil then
+    module:log('error', 'xxl rooms not enabled missing main_muc config');
+    return ;
+end
+
+function table_length(T)
+    local count = 0
+    for _ in pairs(T) do count = count + 1 end
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_xxl_conference.lua:46]: At least you didn't need to implement integer addition

File: resources/prosody-plugins/mod_fmuc.lua (hunk 1 above)
Comment (damencho) [reply resources/prosody-plugins/mod_fmuc.lua:1]: Yeah, while setting up the rest at the moment for running several on one machine I will create a md file with instructions and probably I can drop those then.

File: resources/prosody-plugins/mod_fmuc.lua (hunk 4 above)
Comment (damencho) [reply resources/prosody-plugins/mod_fmuc.lua:74]: From muc itself, I think we use it and in other places ...

File: resources/prosody-plugins/mod_fmuc.lua (hunk 3 above)
Comment (damencho) [reply resources/prosody-plugins/mod_fmuc.lua:30]: Nope, as jicofo is the first one to join, only jicofo can create the rooms.

File: resources/prosody-plugins/mod_xxl_conference.lua (hunk 6 above)
Comment (damencho) [reply resources/prosody-plugins/mod_xxl_conference.lua:46]: And this seems unused ...

**Review by bgrozev — APPROVED** [review]

**Pull request merged**